### PR TITLE
[AIRFLOW-6881] Bulk fetch DAGRun for create_dag_run

### PR DESCRIPTION
**Pull request by mik-laj**
Another performance optimization.
When I have following DAG file 
```python
from datetime import timedelta

from airflow.models import DAG
from airflow.operators.bash_operator import BashOperator
from airflow.operators.dummy_operator import DummyOperator
from airflow.utils.dates import days_ago

args = {
    'owner': 'airflow',
    'start_date': days_ago(3),
}

def create_dag(dag_number):
    dag = DAG(
        dag_id=f'perf_50_dag_dummy_tasks_{dag_number}_of_50', default_args=args,
        schedule_interval="@once",
        dagrun_timeout=timedelta(minutes=60),
        is_paused_upon_creation=False,
    )

    for j in range(1, 10):
        DummyOperator(
            task_id='task_{}_of_5'.format(j),
            dag=dag
        )

    return dag

for i in range(1, 200):
    globals()[f"dag_{i}"] = create_dag(i)
```
I ran the following code:
```python
from airflow.jobs.scheduler_job import DagFileProcessor

processor = DagFileProcessor([], log)
processor.process_file(DAG_FILE, None, pickle_dags=False)
```
I got the following values.
**Before:**
Query count:  2589
Average time 7980.117 ms
**After:**
Query count:  2390
Average time: 7261.959 ms
**Diff:**
Query count: -199 (-7%)
Average time: -719 ms (-9%)

Thanks for support to @evgenyshulman from Databand!

---
Issue link: [AIRFLOW-6881](https://issues.apache.org/jira/browse/AIRFLOW-6881)

Make sure to mark the boxes below before creating PR: [x]

- [X] Description above provides context of the change
- [X] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by tooptoop4** [comment]
can u get perf timings before/after for dags with 500 tasks in them? with these types of changes there could be regression with different dag shapes

**Comment by mik-laj** [comment]
@tooptoop4 The number of tasks does not affect this, but I will also check this case and give values ..

**Comment by mik-laj** [comment]
```
from datetime import timedelta

from airflow.models import DAG
from airflow.operators.dummy_operator import DummyOperator
from airflow.utils.dates import days_ago

args = {
    'owner': 'airflow',
    'start_date': days_ago(3),
}

def create_dag(dag_number):
    dag = DAG(
        dag_id=f'perf_50_dag_dummy_tasks_{dag_number}_of_50', default_args=args,
        schedule_interval="@once",
        dagrun_timeout=timedelta(minutes=60),
        is_paused_upon_creation=False,
    )

    for j in range(1, 100):
        DummyOperator(
            task_id='task_{}_of_5'.format(j),
            dag=dag
        )

    return dag

for i in range(1, 20):
    globals()[f"dag_{i}"] = create_dag(i)

```
Before:
Query count:  249
Retry took 1333.755 ms

After:
Query count:  230
Retry took 1267.842 ms

**Review by ashb — APPROVED** [review]
This one looks okay.

File: airflow/jobs/scheduler_job.py [hunk 1]
```diff
@@ -699,23 +734,30 @@ def _process_dags(self, dagbag, dags, tis_out):
                 self.log.info("Not processing DAG %s since it's paused", dag.dag_id)
                 continue
 
-            self.log.info("Processing %s", dag.dag_id)
+            dag_id = dag.dag_id
+            self.log.info("Processing %s", dag_id)
+            dag_runs_for_dag = dag_runs_by_dag_id[dag_id] if dag_id in dag_runs_by_dag_id else []
```
Comment (saguziel) [inline airflow/jobs/scheduler_job.py]: maybe just dag_runs_by_dag_id.get(dag_id, [])

File: airflow/models/dagrun.py [hunk 2]
```diff
@@ -125,14 +125,16 @@ def refresh_from_db(self, session=None):
 
     @staticmethod
     @provide_session
-    def find(dag_id=None, run_id=None, execution_date=None,
+    def find(dag_id=None, dag_ids=None, run_id=None, execution_date=None,
```
Comment (Fokko) [inline airflow/models/dagrun.py]: Similar to the comment of @kaxil and @ashb earlier, I think we should combine the `dag_id` and `dag_ids` 👍

Comment (Fokko) [reply airflow/models/dagrun.py]: Can we add type annotations as well? :)

File: airflow/jobs/scheduler_job.py [hunk 3]
```diff
@@ -498,19 +499,26 @@ def update_import_errors(session, dagbag):
...
         Returns DagRun if one is scheduled. Otherwise returns None.
         """
         if dag.schedule_interval and conf.getboolean('scheduler', 'USE_JOB_SCHEDULE'):
-            active_runs = DagRun.find(
-                dag_id=dag.dag_id,
-                state=State.RUNNING,
-                external_trigger=False,
-                session=session
-            )
+            if dag_runs is None:
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py:515]: Actually why do we need this branch? This is a method on DagFileProcessor, so not really user facing.

Comment (mik-laj) [reply airflow/jobs/scheduler_job.py:515]: I did this to facilitate testing of this method both in unit tests and in performance tests.

File: airflow/models/dagrun.py (hunk 2 above)
Comment (mik-laj) [reply airflow/models/dagrun.py]: I will combine dag_id and dag_ids parameter.

Here are only primitive types, so I'll add types annotations.

File: airflow/jobs/scheduler_job.py (hunk 1 above)
Comment (mik-laj) [reply airflow/jobs/scheduler_job.py]: Done

**Pull request merged**